src/OrchardCore/OrchardCore.Roles.Core/RoleTracker.cs

```diff
@@ -0,0 +1,147 @@
+using System.Text.Json;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Caching.Memory;
+using OrchardCore.Infrastructure.Security;
+using OrchardCore.Security;
+
+namespace OrchardCore.Roles.Core;
+
+public class RoleTracker : IRoleTracker
+{
+    private const string _roleTrackerCacheKey = "ROLES_WITH_FULL_ACCESS_TRACKER";
+
+    private readonly IDistributedCache _distributedCache;
+    private readonly IMemoryCache _memoryCache;
```

Comment:
As soon as I saw the interface name I knew I had to be ready for some fun ... Please give me some time to understand this, I am trying to think about how to reason about when we should have this.

Reply:
Not a big fan of the name.

I idea is to keep cache this is memory for ultra-fast retrial. Then distributed cache to cache the the same thing global for any instance that use the object. If there is not cache, then we get the info from the role-manager.

Reply:
I agree with the principle, but the issue is that it makes me think about where we can/should apply it anywhere in OC. This is Document Manager again, and with the same potential flows, meaning that there are technical constraints to this as you are implying that this will work in a distributed way, but there could be other challenges as this is not just read-only (other nodes could also update the same value). This means if a bug happens we'll try to fix it and this will become the same monster as Document Manager.

I need some time to think about this.

Reply:
In .net 9 the new HybridCache would help in such a case and in the document manager.